homeassistant/components/purpleair/config_flow.py

```diff
@@ -0,0 +1,205 @@
+"""Config flow for PurpleAir integration."""
+from __future__ import annotations
+
+from typing import Any
+
+from aiopurpleair import API
+from aiopurpleair.endpoints.sensors import NearbySensorResult
+from aiopurpleair.errors import InvalidApiKeyError, PurpleAirError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_API_KEY, CONF_LATITUDE, CONF_LONGITUDE
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers import aiohttp_client, config_validation as cv
+from homeassistant.helpers.selector import (
+    SelectOptionDict,
+    SelectSelector,
+    SelectSelectorConfig,
+    SelectSelectorMode,
+)
+
+from .const import CONF_SENSOR_INDICES, DOMAIN, LOGGER
+
+CONF_DISTANCE = "distance"
+CONF_NEARBY_SENSOR_OPTIONS = "nearby_sensor_options"
+CONF_SENSOR_INDEX = "sensor_index"
+
+DEFAULT_DISTANCE = 5
+
+API_KEY_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_KEY): cv.string,
+    }
+)
+
+
+@callback
+def async_get_api(hass: HomeAssistant, api_key: str) -> API:
+    """Get an aiopurpleair API object."""
+    session = aiohttp_client.async_get_clientsession(hass)
+    return API(api_key, session=session)
+
+
+@callback
+def async_get_coordinates_schema(hass: HomeAssistant) -> vol.Schema:
+    """Define a schema for searching for sensors near a coordinate pair."""
+    return vol.Schema(
+        {
+            vol.Inclusive(
+                CONF_LATITUDE, "coords", default=hass.config.latitude
+            ): cv.latitude,
+            vol.Inclusive(
+                CONF_LONGITUDE, "coords", default=hass.config.longitude
+            ): cv.longitude,
+            vol.Optional(CONF_DISTANCE, default=DEFAULT_DISTANCE): cv.positive_int,
+        }
+    )
+
+
+@callback
+def async_get_nearby_sensors_schema(options: list[SelectOptionDict]) -> vol.Schema:
+    """Define a schema for selecting a sensor from a list."""
+    return vol.Schema(
+        {
+            vol.Required(CONF_SENSOR_INDEX): SelectSelector(
+                SelectSelectorConfig(options=options, mode=SelectSelectorMode.DROPDOWN)
+            )
+        }
+    )
+
+
+class FlowError(Exception):
+    """Define an exception that indicates a flow error."""
+
```

Comment:
It would be nice if this was named instead of a generic exception, but since you already have named expectation, I'd drop the `FlowError` and trap `InvalidApiKeyError` and `PurpleAirError` directly instead.

Reply:
Maybe you could pass unknown errors as strings so the user gets some more feedback if its not something we thought about ahead of time

example
https://github.com/home-assistant/core/blob/5ef972c5d9fe3f7bd38f034c8f9d0ad6aedba903/homeassistant/components/homekit_controller/config_flow.py#L473

Reply:
This isn't just for unknown errors, though—I'm trying to handle errors in a way that will work identically for both config and options flow.

I suppose I could adjust my validation methods (`async_validate_api_key`, `async_validate_coordinates`) to return a dict of errors...

Reply:
@bdraco What do you think about https://github.com/home-assistant/core/pull/82110/commits/441b5f1bd4a9ad31460b41483c45f1fc2d03a20e? This removes `FlowError` while still allowing me to not replicate exception-handling in that future PR.

Reply:
LGTM 👍